src/webgpu/web_platform/canvas/getCurrentTexture.spec.ts

```diff
@@ -260,3 +277,103 @@ g.test('resize')
     currentTexture = ctx.getCurrentTexture();
     t.expect(prevTexture === currentTexture);
   });
+
+g.test('expiry')
+  .desc(
+    `
+Test automatic WebGPU canvas texture expiry on all canvas types with the following requirements:
+- getCurrentTexture returns the same texture object until the next task:
+  - after previous frame update the rendering
+  - before current frame update the rendering
+  - in a microtask off the current frame task
+- getCurrentTexture returns a new texture object and the old texture object becomes invalid
+  as soon as possible after HTML update the rendering.
+
+TODO: make [1] a warning instead of error.
+
+TODO: test more canvas types, and ways to update the rendering
+- if on the same thread, expiry happens when the document updates its rendering (window "rPAF") OR transferToImageBitmap is called
+- if on a different thread, expiry happens when the worker updates its rendering (worker "rPAF") OR transferToImageBitmap is called
+- [draw, transferControlToOffscreen, then canvas is displayed] on either {main thread, or transferred to worker}
+- [draw, canvas is displayed, then transferControlToOffscreen] on either {main thread, or transferred to worker}
+- reftests for the above 2 (what gets displayed when the canvas is displayed)
+- with canvas element added to DOM or not (applies to other canvas tests as well)
+  - canvas is added to DOM after being rendered
+  - canvas is already in DOM but becomes visible after being rendered
+  `
+  )
+  .params(u =>
+    u //
+      .combine('canvasType', kAllCanvasTypes)
+      .combine('prevFrameCallsite', ['runInNewCanvasFrame', 'requestAnimationFrame'] as const)
+      .combine('getCurrentTextureAgain', [true, false] as const)
+  )
+  .fn(t => {
+    const { canvasType, prevFrameCallsite, getCurrentTextureAgain } = t.params;
+    const ctx = t.initCanvasContext(t.params.canvasType);
+    // Create a bindGroupLayout to test invalid texture view usage later.
+    const bgl = t.device.createBindGroupLayout({
+      entries: [
+        {
+          binding: 0,
+          visibility: GPUShaderStage.COMPUTE,
+          texture: {},
+        },
+      ],
+    });
+
+    // The fn is called immediately after previous frame updating the rendering.
+    // Polyfill by calling the callback by setTimeout, in the requestAnimationFrame callback (for onscreen canvas)
+    // or after transferToImageBitmap (for offscreen canvas).
+    function runInNewCanvasFrame(fn: () => void) {
+      t.requestNewFrameOnCanvasType(canvasType, ctx, () => {
```

Comment:
[2] I'd remove this abstraction and call requestPostAnimationFrame / transferToImageBitmap directly here.

where we implement requestPostAnimationFrame as a standalone thing (instead of having the rAF and timeout in different places in the code)